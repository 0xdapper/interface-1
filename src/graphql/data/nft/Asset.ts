import { gql, useQuery } from '@apollo/client'
import { parseEther } from 'ethers/lib/utils'
import { GenieAsset, Rarity, SellOrder } from 'nft/types'

import { NftAssetsFilterInput, NftAssetSortableField } from '../__generated__/types-and-hooks'

const assetQuery = gql`
  query AssetQuery(
    $address: String!
    $orderBy: NftAssetSortableField
    $asc: Boolean
    $filter: NftAssetsFilterInput
    $first: Int
    $after: String
    $last: Int
    $before: String
  ) {
    nftAssets(
      address: $address
      orderBy: $orderBy
      asc: $asc
      filter: $filter
      first: $first
      after: $after
      last: $last
      before: $before
    ) {
      edges {
        node {
          id
          name
          ownerAddress
          image {
            url
          }
          smallImage {
            url
          }
          originalImage {
            url
          }
          tokenId
          description
          animationUrl
          suspiciousFlag
          collection {
            name
            isVerified
            image {
              url
            }
            creator {
              address
              profileImage {
                url
              }
              isVerified
            }
            nftContracts {
              address
              standard
            }
          }
          listings(first: 1) {
            edges {
              node {
                address
                createdAt
                endAt
                id
                maker
                marketplace
                marketplaceUrl
                orderHash
                price {
                  currency
                  value
                }
                quantity
                startAt
                status
                taker
                tokenId
                type
                protocolParameters
              }
              cursor
            }
          }
          rarities {
            provider
            rank
            score
          }
          metadataUrl
        }
      }
      pageInfo {
        endCursor
        hasNextPage
        hasPreviousPage
        startCursor
      }
    }
  }
`

export function useAssetsQuery(
  address: string,
  orderBy: NftAssetSortableField,
  asc: boolean,
  filter: NftAssetsFilterInput,
  first?: number,
  after?: string,
  last?: number,
  before?: string
) {
  const { loading, data, fetchMore } = useQuery(assetQuery, {
    variables: {
      address,
      orderBy,
      asc,
      filter,
      first,
      after,
      last,
      before,
    },
    pollInterval: 999,
  })

  const loadMore = () =>
    fetchMore({
      variables: {
        after: data?.nftAssets?.pageInfo?.endCursor,
      },
    })

  const hasNext = data?.nftAssets?.pageInfo?.hasNextPage

  const assets: GenieAsset[] = data?.nftAssets?.edges?.map((queryAsset: { node: any }) => {
    const asset = queryAsset.node
    const ethPrice = parseEther(
      asset.listings?.edges[0]?.node.price.value?.toLocaleString('fullwide', { useGrouping: false }) ?? '0'
    ).toString()
    return {
      id: asset.id,
      address: asset.collection.nftContracts[0]?.address,
      notForSale: asset.listings?.edges.length === 0,
      collectionName: asset.collection?.name,
      collectionSymbol: asset.collection?.image?.url,
      imageUrl: asset.image?.url,
      animationUrl: asset.animationUrl,
      marketplace: asset.listings?.edges[0]?.node.marketplace.toLowerCase(),
      name: asset.name,
      priceInfo: asset.listings
        ? {
            ETHPrice: ethPrice,
            baseAsset: 'ETH',
            baseDecimals: '18',
            basePrice: ethPrice,
          }
        : undefined,
      susFlag: asset.suspiciousFlag,
      sellorders: asset.listings?.edges.map((listingNode: { node: SellOrder }) => {
        return {
          ...listingNode.node,
          protocolParameters: listingNode.node.protocolParameters
            ? JSON.parse(listingNode.node.protocolParameters.toString())
            : undefined,
        }
      }),
      smallImageUrl: asset.smallImage?.url,
      tokenId: asset.tokenId,
      tokenType: asset.collection.nftContracts[0]?.standard,
      // totalCount?: number, // TODO waiting for BE changes
      collectionIsVerified: asset.collection?.isVerified,
      rarity: {
        primaryProvider: 'Rarity Sniper', // TODO update when backend adds more providers
        providers: asset.rarities.map((rarity: Rarity) => {
          return {
            ...rarity,
            provider: 'Rarity Sniper',
          }
        }),
      },
      owner: asset.ownerAddress,
      creator: {
        profile_img_url: asset.collection?.creator?.profileImage?.url,
        address: asset.collection?.creator?.address,
      },
      metadataUrl: asset.metadataUrl,
    }
  })
  return { assets, loading, loadMore, hasNext }
}
