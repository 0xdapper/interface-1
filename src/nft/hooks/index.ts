export * from './useCart'
export * from './useCollectionFilters'
export * from './useFiltersExpanded'
export * from './useGenieList'
export * from './useIsMobile'
export * from './useLazyEffect'
export * from './useMarketplaceSelect'
export * from './useNFTList'
export * from './useNFTSelect'
export * from './useOnClickOutside'
export * from './useSearchHistory'
export * from './useSelectAsset'
export * from './useSellAsset'
export * from './useSellPageState'
export * from './useSweep'
export * from './useToggle'
export * from './useWalletBalance'
export * from './useWalletCollections'
export * from './useWindowDimensions'
export * from './useWrongNetwork'
