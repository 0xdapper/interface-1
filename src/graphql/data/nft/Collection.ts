import graphql from 'babel-plugin-relay/macro'
import { GenieCollection } from 'nft/types'
import { loadQuery, usePreloadedQuery } from 'react-relay'

import RelayEnvironment from '../RelayEnvironment'
import { CollectionQuery } from './__generated__/CollectionQuery.graphql'

const collectionQuery = graphql`
  query CollectionQuery($address: String!) {
    nftCollections(filter: { addresses: [$address] }) {
      edges {
        cursor
        node {
          bannerImage {
            url
          }
          collectionId
          description
          discordUrl
          homepageUrl
          image {
            url
          }
          instagramName
          isVerified
          name
          numAssets
          twitterName
          nftContracts {
            address
            chain
            name
            standard
            symbol
            totalSupply
          }
          traits {
            name
            values
            stats {
              name
              value
              assets
              listings
            }
          }
          markets(currencies: ETH) {
            floorPrice {
              currency
              value
            }
            owners
            totalVolume {
              value
              currency
            }
            listings
            volume(duration: DAY) {
              value
              currency
            }
            volumePercentChange(duration: DAY) {
              value
              currency
            }
            floorPricePercentChange(duration: DAY) {
              value
              currency
            }
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
        hasPreviousPage
        startCursor
      }
    }
  }
`
export function useCollectionQuery(address: string): GenieCollection | undefined {
  const collectionQueryReference = loadQuery<CollectionQuery>(RelayEnvironment, collectionQuery, {
    address,
  })
  const queryData = usePreloadedQuery<CollectionQuery>(collectionQuery, collectionQueryReference)
  // console.log(queryData)

  const queryCollection = queryData.nftCollections?.edges[0].node
  return {
    // collectionAddress: address,
    address,
    // indexingStatus: string,
    isVerified: queryCollection?.isVerified ?? undefined,
    name: queryCollection?.name ?? undefined,
    description: queryCollection?.description ?? undefined,
    standard: queryCollection?.nftContracts ? queryCollection?.nftContracts[0].standard ?? undefined : undefined,
    bannerImageUrl: queryCollection?.bannerImage?.url ?? undefined,
    floorPrice: queryCollection?.markets ? queryCollection?.markets[0].floorPrice?.value ?? undefined : undefined,
    stats: queryCollection?.markets
      ? {
          num_owners: queryCollection?.markets[0].owners ?? undefined,
          floor_price: queryCollection?.markets[0].floorPrice?.value ?? undefined,
          one_day_volume: queryCollection?.markets[0].volume?.value ?? undefined,
          one_day_change: queryCollection?.markets[0].volumePercentChange?.value ?? undefined,
          one_day_floor_change: queryCollection?.markets[0].floorPricePercentChange?.value ?? undefined,
          banner_image_url: queryCollection?.bannerImage?.url ?? undefined,
          total_supply: queryCollection?.numAssets ?? undefined,
          total_listings: queryCollection?.markets[0].listings ?? undefined,
          total_volume: queryCollection?.markets[0].totalVolume?.value ?? undefined,
        }
      : undefined,
    // symbol: queryCollection.image.url,
    // traits: { // TODO structure traits
    //   trait_type: string
    //   trait_value: string
    //   trait_count: number
    //   floorSellOrder: PriceInfo
    //   floorPrice: number
    // }[],
    // marketplaceCount: { marketplace: string; count: number }[], // TODO structure
    imageUrl: queryCollection?.image?.url ?? undefined,
    twitter: queryCollection?.twitterName ?? undefined,
    instagram: queryCollection?.instagramName ?? undefined,
    discordUrl: queryCollection?.discordUrl ?? undefined,
    externalUrl: queryCollection?.homepageUrl ?? undefined,
    // rarityVerified: boolean, // TODO move check to individual assets
    // isFoundation: boolean, // TODO ask backend how to derive
  }
}
