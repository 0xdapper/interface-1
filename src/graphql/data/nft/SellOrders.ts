import graphql from 'babel-plugin-relay/macro'
import { SellOrder } from 'nft/types'
import { useEffect, useMemo, useState } from 'react'
import { fetchQuery, useRelayEnvironment } from 'react-relay'
import { Subscription } from 'relay-runtime'

import { DetailsSellOrdersQuery } from './__generated__/DetailsSellOrdersQuery.graphql'

const sellOrdersQuery = graphql`
  query SellOrdersQuery($address: String!, $tokenId: String!) {
    nftAssets(address: $address, filter: { tokenIds: [$tokenId] }) {
      edges {
        node {
          listings(first: 10) {
            edges {
              node {
                address
                createdAt
                endAt
                id
                maker
                marketplace
                marketplaceUrl
                orderHash
                price {
                  currency
                  value
                }
                quantity
                startAt
                status
                taker
                tokenId
                type
                protocolParameters
              }
              cursor
            }
          }
        }
      }
    }
  }
`

export function useSellOrdersQuery(
  address: string,
  tokenId: string,
  enabled: boolean
): { sellOrders: SellOrder[] | undefined; isLoading: boolean } {
  const environment = useRelayEnvironment()
  const [sellOrders, setSellOrders] = useState<SellOrder[] | undefined>()
  const [isLoading, setIsLoading] = useState(true)

  useEffect(() => {
    let subscription: Subscription | undefined

    if (enabled) {
      subscription = fetchQuery<DetailsSellOrdersQuery>(environment, sellOrdersQuery, {
        address,
        tokenId,
      }).subscribe({
        next(data) {
          const asset = data.nftAssets?.edges[0]?.node
          const queriedSellOrders = asset?.listings?.edges.map((listingNode) => {
            return {
              ...listingNode.node,
              protocolParameters: listingNode.node.protocolParameters
                ? JSON.parse(listingNode.node.protocolParameters.toString())
                : undefined,
            } as SellOrder
          })
          setSellOrders(queriedSellOrders)
        },
        complete: () => setIsLoading(false),
      })
    }

    return () => {
      if (subscription) subscription.unsubscribe()
    }
  }, [address, enabled, environment, tokenId])

  return useMemo(
    () => ({
      sellOrders,
      isLoading,
    }),
    [sellOrders, isLoading]
  )
}
