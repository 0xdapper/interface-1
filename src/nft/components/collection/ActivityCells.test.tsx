import { TEST_NFT_ACTIVITY_EVENT } from 'test-utils/nft/fixtures'
import { render, screen } from 'test-utils/render'

import { BuyCell } from './ActivityCells'

// TODO: add snapshot matching test when VE works with snapshot testing
describe('BuyCell', () => {
  it('renders add to bag button', async () => {
    const { fragment } = render(
      <BuyCell
        event={TEST_NFT_ACTIVITY_EVENT}
        collectionName="Azuki"
        selectAsset={() => undefined}
        removeAsset={() => undefined}
        itemsInBag={[]}
        cartExpanded={false}
        toggleCart={() => undefined}
        isMobile={false}
        ethPriceInUSD={0}
      />
    )

    expect(fragment).toMatchSnapshot()
    expect(await screen.findByText(/Add to Bag/i)).toBeInTheDocument()
  })
})
