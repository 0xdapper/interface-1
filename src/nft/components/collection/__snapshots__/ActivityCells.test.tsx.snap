// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyCell renders add to bag button 1`] = `undefined`;
